File: 2021/Day6/Next Greater Element 1/Next_greater_element_1.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        mem = defaultdict(lambda: '')
        stack = []

        for num in nums2:
            if stack:
                while stack and stack[-1] < num:
                    mem[stack.pop()] = num
            stack.append(num)

        while stack:
            mem[stack.pop()] = -1

        for i in range(len(nums1)):
            nums1[i] = mem[nums1[i]]

        return nums1

File: 2021/Day6/Next Greater Element 1/Next_greater_element_1.cpp
class Solution
{
public:
    vector<int> nextGreaterElement(vector<int> &nums1, vector<int> &nums2)
    {
        stack<int> stk;
        unordered_map<int, int> mp;

        for (auto &x : nums2)
        {
            while (!stk.empty() && stk.top() < x)
            {
                mp[stk.top()] = x;
                stk.pop();
            }
            stk.push(x);
        }

        vector<int> ans;
        for (auto &x : nums1)
            ans.push_back(mp.count(x) ? mp[x] : -1);
        return ans;
    }
};